src/lib/ares_options.c

```diff
@@ -474,6 +487,32 @@ ares_status_t ares__init_by_options(ares_channel_t            *channel,
     }
   }
 
+  /* Set the thresholds for server failure/recovery if present. */
+  if (optmask & ARES_OPT_SERVER_FAIL) {
+    if (options->server_failure_threshold <= 0) {
+      optmask &= ~(ARES_OPT_SERVER_FAIL);
+    } else {
+      channel->server_failure_threshold =
+        (size_t)options->server_failure_threshold;
+    }
+  }
+  if (optmask & ARES_OPT_SERVER_RECOVER) {
+    if (options->server_recovery_threshold <= 0) {
+      optmask &= ~(ARES_OPT_SERVER_RECOVER);
+    } else {
+      channel->server_recovery_threshold =
+        (size_t)options->server_recovery_threshold;
+    }
+  }
+  if (optmask & ARES_OPT_PRIORITY_SERVER) {
+    if (options->priority_server_chance <= 0) {
+      optmask &= ~(ARES_OPT_PRIORITY_SERVER);
+    } else {
+      channel->priority_server_chance =
+        (size_t)options->priority_server_chance;
+    }
```

Comment:
Is there maybe a maximum value we should enforce here?  due to grabbing a random value in the range of 0-255, and using a % to check for probability, it seems like there should be some range restrictions for it to be usable.  That said, I guess you have a 1 in 256 chance of the random value itself being 0 so anything above that range would still every once in a while "hit".